### PR TITLE
Add AO-screened XC evaluation

### PR DESCRIPTION
**Pull request by JensWehner**
## Summary

- Add blockwise CPU and GPU AO evaluation with active-AO screening.
- Add custom autograd support using the analytic linear-feature pullback.
- Add spatial grid decomposition, permutations, and reusable layout caching.
- Add dense and screened XC integration paths for:
  - electron counts
  - energies
  - XC potentials
  - response and Hessian-vector products
- Add atom-major model chunking and screened-evaluation memory estimators.
- Integrate `SkalaGrids` with CPU and GPU4PySCF classes.
- Route `SkalaNumInt` through the new XC integrator.
- Add dense-versus-screened correctness coverage for RKS, UKS, response, and complete nuclear gradients.
- Add tests for AO slicing/scattering, grid caching, chunking, memory limits, degenerate spatial layouts, and CPU/GPU agreement.

### CODE REVIEW

File: src/skala/pyscf/model_chunking.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+            )
+            max_atoms_per_grid_size[grid_size] = 1
+
+    return ModelFeatureChunker(
+        atom_major_raw_features=atom_major_raw_features,
+        grid_features=grid_features,
+        feature_function=feature_function,
+        chunk_layouts=_make_atom_grid_chunks(
+            sorted_atomic_grid_sizes, max_atoms_per_grid_size
+        ),
```
[the remark this answers is not in the document]
Comment (JensWehner) [reply src/skala/pyscf/model_chunking.py]: atoms in the model do not interact so the function is atom additive.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -102,12 +117,93 @@ jobs:
...
+          python=3.12
+          pyscf=2.13.1
+
+    - name: Install package in development mode
+      run: |
+        pip install -e . --no-deps
+      shell: micromamba-shell {0}
+
+    - name: Run profiling tests
+      run: pytest -v -m profiling tests/test_ao_screening_benchmark.py
```
Comment (tvogels) [inline .github/workflows/test.yml]: Should we just run all tests marked with 'profiling' rather than only this one? I would expect that behavior into the future if we have more profiling tests.

File: tests/utils.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                else mol[feature].square()
+            )
+            for feature in self._quadratic_features
+        ]
+        energy_density = torch.stack(quadratic_terms).sum(dim=0)
+        return (energy_density * grid_weights).sum()
+
+
+@contextmanager
+def patch_ao_screening(
```
Comment (tvogels) [inline tests/utils.py]: naming suggestion: force_ao_screening_mode? I had to look up the function to understand what the context manager does.

File: .github/workflows/test.yml (hunk 2 above)
Comment (tvogels) [inline .github/workflows/test.yml]: If you run all these benchmarks in the CI, I think it would be valuable to also save the raw numbers in CSV or JSON as a CI asset. That will also let us detect performance regressions later. Correct me if I missed this, but this is not there yet, right?

File: src/skala/gpu4pyscf/grids.py [hunk 4]
```diff
@@ -3,17 +3,71 @@
...
+    SpatialGridLayout,
+    prepare_spatial_grid_layout,
+)
+
 LOG = getLogger(__name__)
 
 
-class UnsortableGrids(gen_grid.Grids):  # type: ignore
+class SkalaGrids(gen_grid.Grids):  # type: ignore
+    """GPU4PySCF grids with atom-major ordering and Skala layout caching."""
```
Comment (tvogels) [inline src/skala/gpu4pyscf/grids.py]:
```suggestion
    """
    GPU4PySCF grids with atom-major ordering and Skala layout caching.
    
    Skala 1.1 requires its input grids to be sorted by atom. 
    For efficient computation of the ao features that go into the skala model, however, it's best if the grid points are spatially localized.
    This grid allows switching between the two modes by caching the permutation between these two orderings.
    """
```

File: pyproject.toml [hunk 5]
```diff
@@ -59,11 +62,10 @@ python_version = "3.12"
 exclude = ["third_party/"]
 disable_error_code = ["no-any-return"]
 
-# torch.autograd.Function uses dynamic attributes on FunctionCtx (ctx.save_for_backward pattern)
-# and Function.apply() is untyped. These are fundamental to PyTorch's autograd API.
+# torch.autograd.Function.apply() is untyped in PyTorch.
 [[tool.mypy.overrides]]
-module = "skala.pyscf.features"
-disable_error_code = ["no-any-return", "attr-defined", "no-untyped-call"]
+module = ["skala.pyscf.ao_evaluation", "skala.pyscf.spatial_grid_layout"]
```
Comment (tvogels) [inline pyproject.toml]: what does this do?

File: src/skala/pyscf/model_chunking.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    )
+    grid_indices = (
+        original_starts.index_select(0, point_atom_indices)
+        + torch.arange(total_grid_points, device=atomic_grid_sizes.device)
+        - point_sorted_starts
+    )
+    return AtomGridOrder(atom_indices=atom_indices, grid_indices=grid_indices)
+
+
+class ModelFeatureChunk(NamedTuple):
```
Comment (tvogels) [inline src/skala/pyscf/model_chunking.py:113]: Shall we use dataclasses whenever applicable for consistency?

File: src/skala/pyscf/model_chunking.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
+# SPDX-License-Identifier: MIT
+
+"""Build atom-aligned model feature chunks from globally evaluated raw features.
+
+This module controls how many complete atomic grids are fed through the functional
+model at once. Atomic grids are never split because model features may depend on
+atom-local shapes and coordinates.
+"""
```
Comment (tvogels) [inline src/skala/pyscf/model_chunking.py:8]: Can you extend the docstring with a small doctest explaining how the public api of this module is meant to be used?

File: tests/test_xc_integrator.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    response = integrator.gen_response(mol, grids, dm.detach().clone())
+
+    assert isinstance(result, XCResult)
+    torch.testing.assert_close(result.electron_count, dm.new_tensor(16.0))
+    torch.testing.assert_close(result.energy, dm.new_tensor(128.0))
+    torch.testing.assert_close(result.potential, torch.full_like(dm, 32.0))
+    torch.testing.assert_close(response(torch.ones_like(dm)), torch.full_like(dm, 16.0))
+
+
+def test_xc_integrator_requires_skala_grids_for_density() -> None:
```
Comment (tvogels) [inline tests/test_xc_integrator.py]: This test doesn't call .density, so name is probably not right.

File: src/skala/gpu4pyscf/grids.py [hunk 9]
```diff
@@ -3,17 +3,71 @@
...
+        super().__setattr__("_spatial_grid_layout", None)
+        super().__setattr__("alignment", 1)
+        super().__setattr__("_initializing", False)
+
+    def __setattr__(self, key: str, value: Any) -> None:
+        if (
+            key == "alignment"
+            and value != 1
+            and not getattr(self, "_initializing", False)
+        ):
```
Comment (tvogels) [inline src/skala/gpu4pyscf/grids.py:48]: This doesn't intercept changes like `grid.coords[:] += 0.25`. That's really tricky to avoid in Python. Maybe we should just document it as a limitation? Or do you see a way to protect against it?

File: src/skala/pyscf/xc_integrator.py [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+        grids: Grid,
+        dm0: Tensor,
+        max_memory: int = 2000,
+        safety_fraction: float | None = None,
+    ) -> Callable[[Tensor], Tensor]:
+        """Build an XC-only Hessian-vector product callable."""
+        self._validate_device(dm0)
+        _assert_skala_grid(grids, self.device)
+        if self.feature_spec.supports_spatial_decomposition and _should_screen_aos(mol):
+            return self._gen_response_screened(
```
Comment (tvogels) [inline src/skala/pyscf/xc_integrator.py:123]: These closures capture grid-dependent caches indefinitely. Maybe there is a way to protect against outdated data? Some revision counter? Maybe there is something simpler.

**Review by tvogels — COMMENTED** [review]
Looks great. Main thing I wonder if we can improve is the state handling in the new grid class. There are some, slightly contrived, things you can do to invalidate the caches, like

```python
response = integrator.gen_response(mol, grids, dm)
grids.reset(); grids.level = 1; grids.build()
response(direction)  # old grid
```

or 

```python
grids.build()
screened_energy()
grids.coords[:] += 0.25
screened_energy()  # still old
dense_energy()
```

File: tests/utils.py (hunk 3 above)
Comment (JensWehner) [reply tests/utils.py]: much better thanks.

File: .github/workflows/test.yml (hunk 2 above)
Comment (JensWehner) [reply .github/workflows/test.yml]: Yeah, agreed. So this currently only runs the profiling but not the benchmarks, we could add full benchmarks with validation against reference values. Unfortunately it is super hard to not have them flaky on the CI, so yes I want that, but I think that is a bigger undertaking, maybe even using custom runners.

File: pyproject.toml (hunk 5 above)
Comment (JensWehner) [reply pyproject.toml]: it disables some mypy errors for these files. Do you prefer type ignores in the files instead, maybe more selective.

File: src/skala/pyscf/model_chunking.py (hunk 6 above)
Comment (JensWehner) [reply src/skala/pyscf/model_chunking.py:113]: That is a good question, 

I personally like ManedTuple more because a) faster(not relevant here) b) immutable by default c) can be deconstructed using `a, b = MoldeFeatureChunk` but I do not have a super strong preference.

File: src/skala/gpu4pyscf/grids.py (hunk 9 above)
Comment (JensWehner) [reply src/skala/gpu4pyscf/grids.py:48]: Good question, I am not sure what to do here. Pyscf offers so many ways to change objects that they are basically not encapsulated. I could make these readonly so that they fail with a runtime error?

File: src/skala/pyscf/xc_integrator.py (hunk 10 above)
Comment (JensWehner) [reply src/skala/pyscf/xc_integrator.py:123]: Yeah so in my mind, pyscf intends these "operators" to be short lived when returned, but yeah if someone saves them and then it is modified, weird things will happen.

File: src/skala/pyscf/model_chunking.py (hunk 6 above)
Comment (tvogels) [reply src/skala/pyscf/model_chunking.py:113]: > immutable by default

we should use frozen dataclasses if we do

> can be deconstructed using `a, b = ModelFeatureChunk`

to me this looks like a downside if there is no meaningful ordering in the data. It also allows things like `for x in namedtuple` which doesn't make sense in my opinion.

But I don't have strong preferences for one or the other. The request is guided more by a desire for consistency.

File: pyproject.toml (hunk 5 above)
Comment (tvogels) [reply pyproject.toml]: Ah i just didn't know what it did. Up to you. Maybe type ignore is clearer to me.

Comment (JensWehner) [reply pyproject.toml]: I agree, will change. I am busy with a very flaky CI at the moment. Sorry.

File: src/skala/pyscf/model_chunking.py (hunk 6 above)
Comment (JensWehner) [reply src/skala/pyscf/model_chunking.py:113]: okay will change.

Comment (JensWehner) [reply src/skala/pyscf/model_chunking.py:113]: actually we use NamedTuple in the repo already in the irreps code. https://github.com/search?q=repo%3Amicrosoft%2Fskala+NamedTuple&type=code

File: tests/test_xc_integrator.py (hunk 8 above)
Comment (JensWehner) [reply tests/test_xc_integrator.py]: yep, it should just test, that the XCIntegrator needs a SkalaGrid not something else, clarified name.

File: src/skala/pyscf/xc_integrator.py (hunk 10 above)
Comment (JensWehner) [reply src/skala/pyscf/xc_integrator.py:123]: added docstrings.

File: src/skala/gpu4pyscf/grids.py (hunk 9 above)
Comment (JensWehner) [reply src/skala/gpu4pyscf/grids.py:48]: so for cpu I could restrain that a bit more, added documentation.

File: src/skala/pyscf/model_chunking.py (hunk 7 above)
Comment (JensWehner) [reply src/skala/pyscf/model_chunking.py:8]: Yeah, there was too much going on there, There is one clean class now with a docstring and a much reduced interface.

Comment (JensWehner) [reply src/skala/pyscf/model_chunking.py:8]: I feel it needs a lot less explanation now, but I can add examples if you need.

File: pyproject.toml (hunk 5 above)
Comment (JensWehner) [reply pyproject.toml]: changed the typing contract.

File: .github/workflows/test.yml (hunk 2 above)
Comment (JensWehner) [reply .github/workflows/test.yml]: done

File: tests/utils.py (hunk 3 above)
Comment (JensWehner) [reply tests/utils.py]: renamed.

File: src/skala/pyscf/model_chunking.py (hunk 6 above)
Comment (tvogels) [reply src/skala/pyscf/model_chunking.py:113]: it's because torch compile doesn't work with anything else :)

**Review by tvogels — DISMISSED** [review]
Great, thanks!

File: src/skala/pyscf/model_chunking.py (hunk 6 above)
Comment (JensWehner) [reply src/skala/pyscf/model_chunking.py:113]: ah I see, I had no idea. I can change them to dataclasses then, if you prefer.

**Review by tvogels — APPROVED** [review]

**Review by awvwgk — APPROVED** [review]
Looks good to me
